filters/args.go

```diff
@@ -0,0 +1,295 @@
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.String()
+}
+
+func (a *FilterArgs) Strings() (result []string) {
+	if a.pos > len(a.args) {
+		return nil
+	}
+	hasErr := false
+	for _, x := range a.args[a.pos:] {
+		a.pos++
+		if s, err := StringArg(x); err == nil {
+			result = append(result, s)
+		} else {
+			a.error(err)
+			hasErr = true
+		}
+	}
+	if hasErr {
+		return nil
+	}
+	return
+}
+
+func (a *FilterArgs) Float64() (_ float64) {
+	if x, ok := a.next(); ok {
+		if f, err := Float64Arg(x); err == nil {
+			return f
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+func (a *FilterArgs) OptionalFloat64(defaultValue float64) float64 {
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.Float64()
+}
+
+func (a *FilterArgs) Int64() (_ int64) {
+	if x, ok := a.next(); ok {
+		if i, err := Int64Arg(x); err == nil {
+			return i
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+func (a *FilterArgs) OptionalInt64(defaultValue int64) int64 {
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.Int64()
+}
+
+func (a *FilterArgs) Int() (_ int) {
+	if x, ok := a.next(); ok {
+		if i, err := IntArg(x); err == nil {
+			return i
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+func (a *FilterArgs) OptionalInt(defaultValue int) int {
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.Int()
+}
+
+func (a *FilterArgs) Duration() (_ time.Duration) {
+	if x, ok := a.next(); ok {
+		if d, err := DurationArg(x); err == nil {
+			return d
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+func (a *FilterArgs) OptionalDuration(defaultValue time.Duration) time.Duration {
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.Duration()
+}
+
+// introduced for backwards compatibility, use Duration
+func (a *FilterArgs) DurationOrMilliseconds() (_ time.Duration) {
+	if x, ok := a.next(); ok {
+		if d, err := DurationOrNumberArg(x, time.Millisecond); err == nil {
+			return d
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+// introduced for backwards compatibility, use OptionalDuration
+func (a *FilterArgs) OptionalDurationOrMilliseconds(defaultValue time.Duration) time.Duration {
+	if a.pos >= len(a.args) {
+		return defaultValue
+	}
+	return a.DurationOrMilliseconds()
+}
+
+// introduced for backwards compatibility, use Duration
+func (a *FilterArgs) DurationOrSeconds() (_ time.Duration) {
+	if x, ok := a.next(); ok {
+		if d, err := DurationOrNumberArg(x, time.Second); err == nil {
+			return d
+		} else {
+			a.error(err)
+		}
+	}
+	return
+}
+
+func (a *FilterArgs) Err() error {
+	var errs []string
+	if a.pos != len(a.args) {
+		if a.pos == 1 {
+			errs = append(errs, "expects 1 argument")
+		} else {
+			errs = append(errs, fmt.Sprintf("expects %d arguments", a.pos))
+		}
+	}
+	for _, err := range a.errs {
+		errs = append(errs, err.Error())
+	}
+
+	if len(errs) == 0 {
+		return nil
+	} else {
+		return errors.New(strings.Join(errs, ", "))
```

Comment:
hmm that does not work well with errors.Is https://pkg.go.dev/errors, but I also have no better idea right now.
Maybe with https://pkg.go.dev/errors#As to have a generic FilterArgError?

Reply:
I asked in slack and got pointed to https://github.com/hashicorp/go-multierror

Reply:
Yes, but do we care about? I think the returned error should be treated as a single untyped error. The goal here is to have detailed message. The go-multierror variant would look like this https://github.com/zalando/skipper/tree/filter-args-go-multierror

Reply:
I am thinking about library users, that might want to check `error.Is`, so maybe my first idea with creating FilterArgError type that has the error string that you concatenated. Maybe it's also too complicated and not really useful.